google/cloud/log.h

```diff
@@ -0,0 +1,399 @@
+ * the log level is disabled at compile-time. For compile-time disabled log
+ * levels the compiler should be able to determine that the loop will not
+ * execute at all, and completely eliminate the code (we verified this using
+ * godbolt.org with modern GCC and Clang versions).
+ *
+ * Note the use of fully-qualified class names, including the initial `::`, this
+ * is to prevent problems if anybody uses this macro in a context where `Logger`
+ * is defined by the enclosing namespaces.
+ */
+#define GOOGLE_CLOUD_CPP_LOG_I(level, sink)                                \
+  for (auto GOOGLE_CLOUD_CPP_LOGGER_IDENTIFIER = ::google::cloud::Logger<  \
+           ::google::cloud::LogSink::CompileTimeEnabled(                   \
+               ::google::cloud::Severity::level)>(                         \
+           ::google::cloud::Severity::level, __func__, __FILE__, __LINE__, \
+           sink);                                                          \
+       GOOGLE_CLOUD_CPP_LOGGER_IDENTIFIER.enabled();                       \
+       GOOGLE_CLOUD_CPP_LOGGER_IDENTIFIER.LogTo(sink))                     \
+  GOOGLE_CLOUD_CPP_LOGGER_IDENTIFIER.Stream()
+
+/**
+ * Log a message with the Google Cloud Platform C++ Libraries logging framework.
+ *
+ *
+ */
+#define GCP_LOG(level) GOOGLE_CLOUD_CPP_LOG_I(level, LogSink::Instance())
+
+#ifndef GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED
+#define GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED NOTICE
+#endif  // GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED
+
+/**
+ * Define the severity levels for Google Cloud Platform C++ Libraries logging.
+ *
+ * These are modelled after the severity level in syslog(1) and many derived
+ * tools.
+ *
+ * We force the enum to be represented as an `int` because we will store the
+ * values in an `std::atomic<>` and the implementations usually optimize
+ * `std::atomic<int>` but not `std::atomic<Foo>`
+ */
+enum class Severity : int {
+  /// Use this level for messages that indicate the code is entering and leaving
+  /// functions.
+  TRACE,
+  /// Use this level for debug messages that should not be present in
+  /// production.
+  DEBUG,
+  /// Informational messages, such as normal progress.
+  INFO,
+  /// Informational messages, such as unusual, but expected conditions.
+  NOTICE,
+  /// An indication of problems, users may need to take action.
+  WARNING,
+  /// An error has been detected.  Do not use for normal conditions, such as
+  /// remote servers disconnecting.
+  ERROR,
+  /// The system is in a critical state, such as running out of local resources.
+  CRITICAL,
+  /// The system is at risk of immediate failure.
+  ALERT,
+  /// The system is about to crash or terminate.
+  FATAL,
+  /// The highest possible severity level.
+  HIGHEST = int(FATAL),
+  /// The lowest possible severity level.
+  LOWEST = int(TRACE),
+  /// The lowest level that is enabled at compile-time.
+  LOWEST_ENABLED = int(GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED),
```

Comment:
What's the value of having levels below the default-enabled level? The user is forced to modify this to enable it in their project, but they're not blocked from it.

Reply:
So we can write code like this:

```C++
void Function() {
  GCP_LOG(TRACE) << "Entering function Function()";
  // do stuff
  GCP_LOG(TRACE) << "Leaving function Function()";
}
```

You only compile with `TRACE` enabled if you are desperate and need to debug something with massive amounts of logging.  Most of the time those are not even there.  I do not recommend sprinkling the code with `TRACE` messages, but I thought we would want to support it.  Happy to change the default value for `GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED`.